### DISCUSSION TITLE
implementing a nanoparticle

### DISCUSSION THREAD

**Discussion by kirstynvalles02** [BLAST-WarpX/warpx#5397] [Q&A] [answered]
Greetings!

I am attempting to implement a nanoparticle in a specific spot of my simulation, however i am unsure of how to actually implement one into an imput file. any help would be greatly appreciated!

**Comment by aeriforme** [comment, marked as answer]
Hello! 

If you want to initialize a nanometric sphere in your simulation, you can add a particle species with:
```
<species_name>.profile = parse_density_function
<species_name>.density_function(x,y,z) =  n0 * ( ((x-x0)*(x-x0) + (y-y0)*(y-y0) + (z-z0)*(z-z0)) < R**2)
```
This would add particles in your simulation according to `density_function`, meaning that `n0` would be the number density of that species inside the sphere, `x0,y0,z0` the center and `R` the radius of the sphere.

[This example](https://github.com/ECP-WarpX/WarpX/blob/development/Examples/Tests/electrostatic_sphere/inputs_base_3d) could be of help. 


Otherwise, if what you want is to initialize a single macro particle in a given location, you can use `<species_name>.injection_style = SingleParticle`. In this case, you can check out [this other example](https://github.com/ECP-WarpX/WarpX/blob/development/Examples/Tests/particle_pusher/inputs_test_3d_particle_pusher)

You can find more details about particle initialization [here](https://warpx.readthedocs.io/en/22.11/usage/parameters.html#particle-initialization).


Let me know if this answers your question! :)

**Comment by kirstynvalles02** [comment]
youre honestly amazing! thank you so much!!

may i keep this chat open in case i have further questions??

**Reply by aeriforme** [reply]
I transformed the issue into a discussion, which is more appropriate for questions and...discussions

**Comment by kirstynvalles02** [comment]
hello!

quick question-- would i need a different parameter to declare it as a gold nanoparticle?

**Reply by aeriforme** [reply]
For each species, either its type or its charge and mass must be defined. 
The easiest would be: `<species_name>.species_type = "gold"`.
Note this: 

> When only the name of an element is specified, the mass is a weighted average of the masses of the stable isotopes.


You can look all this up [here](https://warpx.readthedocs.io/en/latest/usage/parameters.html#particle-initialization).

**Comment by kirstynvalles02** [comment]
apologies i could not access this thread a while back, however i did try and run it but i keep getting dimension errors. I will attach the input file and a copy of the error below:
############################################
[sim4nanoinput.txt](https://github.com/user-attachments/files/17548604/sim4nanoinput.txt)
############################################
![image](https://github.com/user-attachments/assets/728e74df-1cb2-48e4-ad13-cb67ed386e03)

**Reply by aeriforme** [reply]
Those are warnings, not errors.
However, they suggest that your species are actually empty.

Please note that when I mentioned `<species_name>`, 
I was using a placeholder for the actual species names that you want to use (in this case `ele` and `ion`).
